pkg/template/jsonpath.go

```diff
@@ -0,0 +1,155 @@
+			return "", err
+		}
+	}
+
+	return buf.String(), nil
+}
+
+// PrintResults writes the results into writer
+// This is a slightly modified copy of the original
+// j.PrintResults from k8s.io/client-go/util/jsonpath/jsonpath.go
+// in that it uses calls `textValue()` for instead of `evalToText`
+// This is a workaround for kubernetes/kubernetes#16707
+func printResults(wr io.Writer, results []reflect.Value) error {
+	for i, r := range results {
+		text, err := textValue(r)
+		if err != nil {
+			return err
+		}
+		if i != len(results)-1 {
+			text = append(text, ' ')
+		}
+		if _, err = wr.Write(text); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// textValue translates reflect value to corresponding text
+// If the value if an array or map, it returns a JSON representation
+// of the value (as opposed to the internal go representation of the value)
+// Otherwise, the text value is from the `evalToText` function, originally from
+// k8s.io/client-go/util/jsonpath/jsonpath.go
+func textValue(v reflect.Value) ([]byte, error) {
+	t := reflect.TypeOf(v.Interface())
+	// special case for null values in JSON; evalToText() returns <nil> here
+	if t == nil {
+		return []byte("null"), nil
+	}
+
+	switch t.Kind() {
+	// evalToText() returns <map> ....; return JSON string instead.
+	case reflect.Map, reflect.Slice:
+		text, err := json.Marshal(v.Interface())
+		if err != nil {
+			return nil, err
+		}
+		return text, nil
+	default:
+		return evalToText(v)
+	}
+}
+
+// evalToText translates reflect value to corresponding text
+// This is a unmodified copy of j.evalToText from k8s.io/client-go/util/jsonpath/jsonpath.go
+func evalToText(v reflect.Value) ([]byte, error) {
+	iface, ok := template.PrintableValue(v)
+	if !ok {
+		return nil, fmt.Errorf("can't print type %s", v.Type())
+	}
+	var buffer bytes.Buffer
+	fmt.Fprint(&buffer, iface)
+	return buffer.Bytes(), nil
+}
+
+
+// TektonJSONPathExpression returns a valid JSONPath expression. It accepts
+// a "RelaxedJSONPath" expression that is wrapped in the Tekton variable
+// interpolation syntax i.e. $(). RelaxedJSONPath expressions can optionally
+// omit the leading curly braces '{}' and '.'
+func TektonJSONPathExpression(expr string) (string, error) {
+	if !isTektonExpr(expr) {
+		return "", errors.New("Expression not wrapped in $()")
+	}
+	return relaxedJSONPathExpression(unwrwapTektonExpr(expr))
+}
+
+// RelaxedJSONPathExpression attempts to be flexible with JSONPath expressions, it accepts:
+//   * metadata.name (no leading '.' or curly braces '{...}'
+//   * {metadata.name} (no leading '.')
+//   * .metadata.name (no curly braces '{...}')
+//   * {.metadata.name} (complete expression)
+// And transforms them all into a valid jsonpath expression:
+//   {.metadata.name}
+func relaxedJSONPathExpression(pathExpression string) (string, error) {
+	if len(pathExpression) == 0 {
+		return pathExpression, nil
+	}
+	submatches := jsonRegexp.FindStringSubmatch(pathExpression)
```

Comment:
Mostly because I don't fully grok the regex being used here, but what exactly are we trying to extract here? (might be helpful to provide an example).

Reply:
The function docstring and the substring error has some examples of what the regex accepts but basically it tries to extract a substring - so:
`{blah}` or `{.blah}` or `.blah` or `blah`

The substring however cannot contain curly braces so `bla{h` won't match